SAK-25099 - Anonymous topics in forums

bbailla2 opened the pull request:
An anonymous feature for forums.

Overview:
In anonymous contexts, users will see 6 uppercase hex characters in place of the authors' names. To help students address each other, and so that instructors may grade in anonymous contexts, each user's anonymous ID will remain consistent within a given site.  When the current user sees their own post, their anonymous ID will be suffixed with "(me)".

The anonymous setting is set by the instructor at the topic level, and it applies to all conversations within that topic. There are two options, and a permission:
Option 1: Posts are anonymous
Option 2: Reveal identities to roles with the "Identify Anonymous Authors" permission below
Permission: Identify Anonymous Authors

Option 2 is available only if option 1 is checked.
The new "Identify Anonymous Authors" permission allows site maintainers to control who will see the revealed identities, but it only applies if option 2 is checked.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -3381,6 +3405,73 @@ else if(getUserId()==null&&this.forumManager.getAnonRole()==true){
+
+		String siteId = ToolManager.getCurrentPlacement().getContext();
+		// Iterate over messages and construct a list of authors
+		List<String> userIds = new ArrayList<String>();
```
I don't think this JIRA is set to be merged into 10.x, so you can use the diamond operator here.

ottenhoff replied:
If the performance of ranks is as bad as it looks, we should probably merge this into 10.x for everyone

bbailla2 replied:
You might be thinking of https://jira.sakaiproject.org/browse/SAK-29936, which seems to be merged

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -4759,14 +4840,23 @@ public String processDfMsgRevisedPost()
+			{
+				revisedInfo += getResourceBundleString(LAST_REVISE_BY);
+				revisedInfo += getUserNameOrEid();
+				revisedInfo += " " + getResourceBundleString(LAST_REVISE_ON);
```
You could condense this String concatenation into one line; not a big deal though.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -4904,9 +4994,18 @@ public String processDfMsgSaveRevisedDraft()
+		      {
+		      	revisedInfo = getResourceBundleString(LAST_REVISE_BY);
+		      	revisedInfo += getUserNameOrEid();
+		      	revisedInfo += " " + getResourceBundleString(LAST_REVISE_ON);
```
Same here.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -5288,15 +5387,65 @@ private void refreshPendingMessages()
+		  //Maps topics to a boolean representing whether anonymousIDs should be displayed within the topic's context
+		  Map<Topic, Boolean> topicUseAnonIdMap = new HashMap<Topic, Boolean>();
+		  // Maps userIds to all the messages that are within an anonymous context
+		  Map<String, List<DiscussionMessageBean>> userIdAnonMessagesMap = new HashMap<String, List<DiscussionMessageBean>>();
```
Diamond operator here and two lines above.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -5288,15 +5387,65 @@ private void refreshPendingMessages()
+			  	List<DiscussionMessageBean> userAnonymousMessages = userIdAnonMessagesMap.get(userId);
+			  	if (userAnonymousMessages == null)
+			  	{
+			  		userAnonymousMessages = new ArrayList<DiscussionMessageBean>();
```
Diamond operator.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -5288,15 +5387,65 @@ private void refreshPendingMessages()
+		  String siteId = ToolManager.getCurrentPlacement().getContext();
+		  // AnonymousManager requires lists (because it needs to sublist into groups of 1000 for Oracle)
+		  // Convert to userIdAnonMessagesMap's keySet into a list
+		  List<String> userIdList = new ArrayList<String>();
```
Diamond operator.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -5578,8 +5727,15 @@ public String processAddCommentToDeniedMsg()
+	  else
+	  {
+	  	sb.append(getResourceBundleString(MOD_COMMENT_TEXT)).append(" ");
+	  	sb.append(UserDirectoryService.getCurrentUser().getDisplayName());
```
You can chain these two .append statements together; again not a big deal.

bbailla2 replied:
I kinda like it like this vertical actually, but I've refactored a few other ones

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -7334,6 +7496,7 @@ else if (target instanceof Topic){
+          mask.put(PermissionLevel.IDENTIFY_ANON_AUTHORS, Boolean.valueOf(permBean.getIdentifyAnonAuthors()));
           mask.put(PermissionLevel.DELETE_OWN, Boolean.valueOf(permBean.getDeleteOwn()));
           mask.put(PermissionLevel.DELETE_ANY, Boolean.valueOf(permBean.getDeleteAny()));
           mask.put(PermissionLevel.REVISE_OWN, Boolean.valueOf(permBean.getReviseOwn()));
```
The above code blocks all look the same. Could this be an opportunity to reduce code bloat by extracting the common pieces into a method?

bbailla2 replied:
Probably. The if statements containing these blocks might not even be necessary (but I'm not certain cause PermissionLevelImpl overrides .equals with a very incoherent implementation). At the very least a method would be nice here, so I'll set one up.

bbailla2 replied:
They're not comparing .equals on PermissionLevelImpl, just String.equals on the name attributes, so I'm pretty whole block will always execute exactly once. Stripping it all out into a single call to a private method!

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ForumsEmailService.java:
```diff
@@ -345,4 +348,30 @@ private boolean isValidEmail(String email) {
 
+	private String getAnonAwareAuthor(Message message)
+	{
+		if (isUseAnonymousId(message.getTopic()))
```
Should the `message` object be null checked here? I'm not sure if it's possible for it to be null.

bbailla2 replied:
It comes from the object's constructor, but a null check wouldn't hurt. Though if it was null, we'd probably throw an IllegalArgumentException. IAE / NPE = nbd.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ForumsEmailService.java:
```diff
@@ -345,4 +348,30 @@ private boolean isValidEmail(String email) {
+		 * This is okay, but if isRevealIDsToRoles is enabled, it would be better to send two separate emails (one revealed, one anonymous) 
+		 * so that the instructor can't match the message up with the UI to discover the anonymous mapping for the author
+		 */
+		return ServerConfigurationService.getBoolean("msgcntr.forums.anonymous.enable", true) && topic.getPostAnonymous();
```
I would put this sakai.property name and default into a static final variable on the class, or in a common place for use. Same with the other place(s) you're checking this property.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionMessageBean.java:
```diff
@@ -449,4 +457,128 @@ public String getAuthorEid(){
+		}
+
+		// get / create the anonId from the AnonymousManager
+		AnonymousManager anonymousManager = (AnonymousManager)ComponentManager.get("org.sakaiproject.api.app.messageforums.AnonymousManager");
```
I think ComponentManager has been updated so you no longer need to cast the return value.

bbailla2 replied:
Yes, but only if you use
<T> T get(Class<T> iface);
So my refactor will switch a bunch of these to:
ComponentManager.get(AnonymousManager.class)
Much cleaner!

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionMessageBean.java:
```diff
@@ -449,4 +457,128 @@ public String getAuthorEid(){
+
+		// get / create the anonId from the AnonymousManager
+		AnonymousManager anonymousManager = (AnonymousManager)ComponentManager.get("org.sakaiproject.api.app.messageforums.AnonymousManager");
+		String siteId = ToolManager.getInstance().getCurrentPlacement().getContext();
```
Covers are deprecated; we should avoid adding them to new code. Please use the `ComponentManager` or Spring injection to get the API.

bbailla2 replied:
Covers are ubiquitous in msgcntr, I'd be in favor of having it all refactored in a separate SAK ticket.

bjones86 replied:
There is an effort to replace Covers and remove them; however we shouldn't be adding more work by introducing new instances of them.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionMessageBean.java:
```diff
@@ -449,4 +457,128 @@ public String getAuthorEid(){
+		this.useAnonymousId = useAnonymousId;
+	}
+
+	public boolean isUseAnonymousId()
```
JavaDocs on these methods would be helpful.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionTopicBean.java:
```diff
@@ -556,6 +558,62 @@ public void setPostFirst(String postFirst)
+   */
+  public void setPostAnonymous(String postAnonymous)
+  {
+    if (LOG.isDebugEnabled())
```
Probably don't need this `if` statement; like the other methods in this class, just fire off the `log.debug`

plukasew replied:
I don't know, I think best practice is to call isDebugEnabled() first because it is a cheap call and debug() isn't.

bbailla2 replied:
I think the best practise is to use the check if you're concatenating Strings

plukasew replied:
Yeah, I think you're right. The benefit is about avoiding any computation involved in creating the string to log. In this case we have some string concat so the guard is warranted.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionTopicBean.java:
```diff
@@ -556,6 +558,62 @@ public void setPostFirst(String postFirst)
+   */
+  public void setRevealIDsToRoles(String revealIDsToRoles)
+  {
+    if (LOG.isDebugEnabled())
```
Same deal here.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionTopicBean.java:
```diff
@@ -615,6 +673,60 @@ public void setTopicPostFirst(Boolean postFirst)
+   */
+  public void setTopicPostAnonymous(Boolean postAnonymous)
+  {
+    if (LOG.isDebugEnabled())
```
And here.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionTopicBean.java:
```diff
@@ -615,6 +673,60 @@ public void setTopicPostFirst(Boolean postFirst)
+   */
+  public void setTopicRevealIDsToRoles(Boolean revealIDsToRoles)
+  {
+    if (LOG.isDebugEnabled())
```
And here too.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/MessageForumStatisticsBean.java:
```diff
@@ -499,6 +532,25 @@ public List getAllUserStatistics(){
 
+		// Determine whether the site is a pure anonymous scenario (Ie. all topics in the site are anonymous)
+		String siteId = ToolManager.getCurrentPlacement().getContext();
+		List<Topic> allTopics = forumManager.getTopicsInSite(siteId);
```
I'm just spit-balling here, but could we leverage the database to return only non-anonymous topics in a new method? That would eliminate having to loop over all topics to find a non-anonymous one. You simply call the method to get all non-anon, size/empty check it and be done with it. Just an idea.

bbailla2 replied:
Wouldn't work in situations where every topic in the site has POST_ANONYMOUS = true, and REVEAL_IDS_TO_ROLES = true, and the user has the identifyAnonAuthors permission.

Though you bring up a good point because most sites will have the "General Discussion" topic which should be revealed 99.9% of the time. So in the common case we avoid this loop and use quick query like:
select count(*) from MFR_TOPIC_T topic (join ... to tie it to the context) where topic.post_anonymous=1 limit 1;

bbailla2 replied:
Yeah, I think it would deliver performance improvements in the common case to prevent the propagation of all the topics in the site from the DB, but it's probably not critical. I'll create a ticket for our OWL roadmap.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/MessageForumStatisticsBean.java:
```diff
@@ -499,6 +532,25 @@ public List getAllUserStatistics(){
+				break;
+			}
+		}
+		Map<String, String> userIdAnonIdMap = new HashMap<String, String>();
```
Diamond operator.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/MessageForumStatisticsBean.java:
```diff
@@ -608,9 +674,31 @@ public List getUserAuthoredStatistics(){
+		{
+			return statistics;
+		}
+		List<UserStatistics> filtered = new ArrayList<UserStatistics>();
```
Diamond operator.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/MessageForumStatisticsBean.java:
```diff
@@ -737,6 +832,19 @@ public List getTopicStatistics(){
+			if (m_displayAnonIds)
+			{
+				String siteId = ToolManager.getCurrentPlacement().getContext();
+				List<String> userIds = new ArrayList<String>();
```
Diamond operator.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/MessageForumStatisticsBean.java:
```diff
@@ -776,6 +890,37 @@ public List getTopicStatistics(){
+	String m_gradeStatisticsAssign = null;
+	String m_gradeStatisticsGroup = null;
+	Comparator m_comparator = null;
+	public List getGradeStatisticsForStatsListByTopic()
```
JavaDocs

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/MessageForumStatisticsBean.java:
```diff
@@ -883,6 +1030,37 @@ public List getUserAuthoredStatistics2(){
+
+		List<Message> filtered = new ArrayList<Message>();
+		// Map topics to value of 'isUseAnonymousId(Topic)' to prevent redundant queries
+		Map<Topic, Boolean> topicToUseAnon = new HashMap<Topic, Boolean>();
```
Diamond operators.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/MessageForumStatisticsBean.java:
```diff
@@ -883,6 +1030,37 @@ public List getUserAuthoredStatistics2(){
+		Map<Topic, Boolean> topicToUseAnon = new HashMap<Topic, Boolean>();
+		for (Message message : messages)
+		{
+			Topic topic = message.getTopic();
```
I've seen that you're null checking the `Topic` object in other places, perhaps here as well? Also, what about the `messages` list being passed into this method?

bbailla2 replied:
In the context of Forums (but not Messages), message.getTopic() is never null. However, accessing the topic's attributes other than the ID may result in LazyInitializationException depending how the message was retrieved. In this case the calling methods use queries that join on the topic table, so we should be safe.

In terms of messages, the calling methods are passing in the result of Query.list(), which is never null. But some null checks etc. here would be good for extra confidence.

bjones86 replied:
Okay cool, thanks for clarifying that.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/MessageForumStatisticsBean.java:
```diff
@@ -1849,43 +1998,61 @@ public int compare(Object item, Object anotherItem){
 		};
 	}
+
+	private static int compareGradesFromStats(DecoratedCompiledMessageStatistics stat1, DecoratedCompiledMessageStatistics stat2)
```
Nice work extracting the above duplicated code into methods. Some JavaDocs would be nice though.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/MessageForumStatisticsBean.java:
```diff
@@ -1971,22 +2146,24 @@ private String getExternalParameterByKey(String parameterId)
+
+		if (m_displayAnonIds)
+		{
+			userName = StringUtils.trimToEmpty(anonymousManager.getAnonId(ToolManager.getCurrentPlacement().getContext(), selectedSiteUserId));
```
Covers are deprecated; ComponentManager or Spring inject the API

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/MessageForumStatisticsBean.java:
```diff
@@ -2063,6 +2252,34 @@ public void setSelectedTopicTitle(String selectedTopicTitle) {
+	private List<MembershipItem> convertNameToAnonIdAndSort(Map<String, MembershipItem> memberMap)
+	{
+		Map<String, String> userIdToAnonIdMap = anonymousManager.getUserIdAnonIdMap(ToolManager.getCurrentPlacement().getContext());
+		List<MembershipItem> list = new ArrayList<MembershipItem>();
```
Diamond operator.

bjones86 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/PermissionBean.java:
```diff
@@ -168,6 +168,20 @@ public void setModeratePostings(boolean moderatePostings)
+  {
+    if (item != null && item.getPermissionLevel() != null 
+        && item.getPermissionLevel().getIdentifyAnonAuthors() != null)
+      return item.getPermissionLevel().getIdentifyAnonAuthors().booleanValue();
```
Braces on single-line conditionals please.

bjones86 commented on msgcntr/messageforums-app/src/webapp/jsp/discussionForum/permissions/permissions_include.jsp:
```diff
@@ -90,6 +90,10 @@
           <h:outputLabel for="moderatePostings"><h:outputText value="#{msgs.perm_moderate_postings}" /></h:outputLabel>
         </h:panelGroup>
+        <h:panelGroup styleClass="checkbox" style="display: #{ForumTool.anonymousEnabled ? '' : 'none'}">
+          <h:selectBooleanCheckbox id="identifyAnonAuthors" value="#{permission.identifyAnonAuthors}" onclick="javascipt:setCorrespondingLevel(this.id);" disabled="#{(not ForumTool.editMode)"/>
```
`javascript:` prefix in the onclick is not actually needed/required.

bbailla2 replied:
Also brace bracket wasn't closed on "#{(not ForumTool.editMode)"

bjones86 commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/AnonymousManagerImpl.java:
```diff
@@ -0,0 +1,198 @@
+import org.springframework.orm.hibernate3.HibernateCallback;
+import org.springframework.orm.hibernate3.support.HibernateDaoSupport;
+
+public class AnonymousManagerImpl extends HibernateDaoSupport implements AnonymousManager
```
JavaDoc on the class would be helpful

bjones86 commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/AnonymousManagerImpl.java:
```diff
@@ -0,0 +1,198 @@
+
+	private final int MAX_IN_CLAUSE_SIZE = 1000;
+
+	public String getAnonId(final String siteId, final String userId)
```
You can use the following JavaDoc on these methods to inherit the docs from the interface:

`/**
* {@inheritDoc}
*/`

bjones86 commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/AnonymousManagerImpl.java:
```diff
@@ -0,0 +1,198 @@
+			return Collections.emptyList();
+		}
+
+		HibernateCallback<List<AnonymousMapping>> hcb = new HibernateCallback<List<AnonymousMapping>>()
```
Diamond operator.

bbailla2 replied:
This is the compilation error that threw me off and lead me to assume that diamond operators didn't compile in msgcntr:
cannot infer type arguments for org.springframework.orm.hibernate3.HibernateCallback<T>
[ERROR] reason: cannot use '<>' with anonymous inner classes

So, no anonymous classes, but I'm applying them everywhere else in the refactor

bjones86 commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/AnonymousManagerImpl.java:
```diff
@@ -0,0 +1,198 @@
+		{
+			public List<AnonymousMapping> doInHibernate(Session session) throws HibernateException, SQLException
+			{
+				List<AnonymousMapping> mappings = new ArrayList<AnonymousMapping>();
```
Diamond operator.

bjones86 commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/AnonymousManagerImpl.java:
```diff
@@ -0,0 +1,198 @@
+	{
+		Map<String, String> anonIdMap = new HashMap<String, String>();
+
+		HibernateCallback<List<AnonymousMapping>> hcb = new HibernateCallback<List<AnonymousMapping>>()
```
Diamond operators.

bjones86 commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/AnonymousManagerImpl.java:
```diff
@@ -0,0 +1,198 @@
+
+	public Map<String, String> getUserIdAnonIdMap(final String siteId, final List<String> userIds)
+	{
+		Map<String, String> anonIdMap = new HashMap<String, String>(userIds.size());
```
Diamond operator.

bjones86 commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/AnonymousManagerImpl.java:
```diff
@@ -0,0 +1,198 @@
+		if (anonIdMap.size() < userIds.size())
+		{
+			// The map doesn't contain all of the users. We'll have to create their anonIds
+			Set<String> notFound = new HashSet<String>();
```
Diamond operator.

bjones86 commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/AnonymousManagerImpl.java:
```diff
@@ -0,0 +1,198 @@
+	{
+		// 6 hex characters is short enough to be recognizable to follow a conversation.
+		// Maximum value = 0xFFFFFF = 16777215
+		String hex = Integer.toHexString( (new Random()).nextInt(16777215) );
```
I would put the maximum value in a static final class member.

bjones86 commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java:
```diff
@@ -1547,6 +1562,57 @@ public Object doInHibernate(Session session) throws HibernateException, SQLExcep
+
+			final String query = anonymousOnly ? "findAnonymousTopicsInSite" : "findTopicsInSite";
+
+			HibernateCallback<List<Topic>> hcb = new HibernateCallback<List<Topic>>()
```
Diamond operator.

buckett replied:
Anonymous inner class again I think.

bjones86 commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java:
```diff
@@ -1547,6 +1562,57 @@ public Object doInHibernate(Session session) throws HibernateException, SQLExcep
+				}
+			};
+
+			List<Topic> topicList = new ArrayList<Topic>();
```
Diamond operator.

bjones86 commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/PermissionLevelManagerImpl.java:
```diff
@@ -534,6 +534,8 @@ else if (customPermName.equals(PermissionLevel.MARK_AS_READ))
     		return permissionLevel.getMarkAsRead();
     	else if (customPermName.equals(PermissionLevel.MODERATE_POSTINGS))
     		return permissionLevel.getModeratePostings();
+    	else if (customPermName.equals(PermissionLevel.IDENTIFY_ANON_AUTHORS))
```
I realize the surrounding code doesn't follow the best practise of using braces for single line conditionals, but we shouldn't introduce more occurrences of a bad practise.

plukasew replied:
+1

bjones86 commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/PermissionLevelManagerImpl.java:
```diff
@@ -913,6 +921,7 @@ private PermissionsMask getDefaultReviewerPermissionsMask() {
+		  mask.put(PermissionLevel.IDENTIFY_ANON_AUTHORS, Boolean.valueOf(false));
 		  mask.put(PermissionLevel.DELETE_OWN, Boolean.valueOf(false));
 		  mask.put(PermissionLevel.DELETE_ANY, Boolean.valueOf(false));
 		  mask.put(PermissionLevel.REVISE_OWN, Boolean.valueOf(false));
```
Like my earlier comment, the above code seems to be duplicated a lot. Opportunity for a refactor?

bbailla2 replied:
This is where the default permissions are set; they differ for each role.
We could create a method like:
setupMask(Mask mask, boolean newForum, boolean newTopic, bolean...)
{
    mask.put(PermissionLevel.NEW_FORUM, newForum);
    mark.put(...
}
and invoke setupMask(mask, true, true, true, ...);
But then it'd be hard to read which boolean is going to which permission when we invoke it.
IMO it's better to lease all of these explicit.

buckett replied:
You can use the builder pattern to make more readable code:

    PermissionMaskBuilder builder = new PermissionMarkBuilder(); // Default everything to false
    PermissionMask mask = builder.addRead().addMarkRead().addDeleteOwn().build();

I'm not suggesting you need todo this here, just suggesting it as a possible coding style solution.

bjones86 commented on msgcntr/messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/PermissionLevelImpl.java:
```diff
@@ -274,6 +286,10 @@ public boolean equals(Object o)
     	return returnValue;
+	bThis = this.getIdentifyAnonAuthors();         
+    returnValue = (bThis == null) ? obj.getIdentifyAnonAuthors() == null : bThis.equals(obj.getIdentifyAnonAuthors());
+    if(!returnValue)
```
Braces on single line conditionals.

bjones86 commented on msgcntr/messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/TopicImpl.java:
```diff
@@ -396,4 +399,24 @@ public void setPostFirst(Boolean postFirst) {
+	}
+
+	public boolean isAnonymousEnabled() {
+		return ServerConfigurationService.getBoolean("msgcntr.forums.anonymous.enable", true);
```
Like my earlier comment, sakai.property and default should probably go in static finals somewhere.

bjones86 commented:
I've done a preliminary code review on this, it seems pretty solid and well documented. I've made some comments about minor things here and there.

That said, this is a significant contribution in terms of both the amount of code added, and the new functionality. I think it would be best to get as many eyes on this as possible. If anyone has the time to do a review and/or test, it would be very beneficial.

plukasew commented on msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages.properties:
```diff
@@ -213,6 +216,9 @@ cdfm_moderate_forum=Moderate topics in forum
 cdfm_postFirst=Require users to post before reading
 cdfm_postFirst_warning=You must post to this topic before you can view any messages
+cdfm_postAnonymous=Posts are anonymous
+cdfm_revealIDsToRoles=Reveal identities to roles with the "Identify Anonymous Authors" permission below
```
This should probably just reference the "perm_identify_anon_authors" message

bbailla2 replied:
Ideally, but the string would have to be assembled in java code somewhere (it's referenced in a jsp). Probably simpler to just leave it typed out in the message bundle.

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -3261,6 +3262,10 @@ private DiscussionTopicBean getDecoratedTopic(DiscussionTopic topic)
 
+		// Determine if we should display authors' userIDs or their anonIDs
+		boolean useAnonymousId = isUseAnonymousId(topic);
+		Map<String, String> userIdAnonIdMap = null;
```
While this will get set below, it is safer to assign Collections.emptyMap() here instead of null.

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -3324,6 +3334,13 @@ private DiscussionTopicBean getDecoratedTopic(DiscussionTopic topic)
     					messageManager);
+    			// Set anonymous attributes on the bean (reduces queries later; improves performance)
+    			decoMsg.setUseAnonymousId(useAnonymousId);
+    			if (userIdAnonIdMap != null)
```
replace with == useAnonymousId

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -3371,6 +3388,13 @@ else if(getUserId()==null&&this.forumManager.getAnonRole()==true){
 						  messageManager);
+				  // Set anonymous attributes on the bean (reduces queries later; improves performance)
+				  decoMsg.setUseAnonymousId(useAnonymousId);
+				  if (userIdAnonIdMap != null)
```
see above

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -3381,6 +3405,73 @@ else if(getUserId()==null&&this.forumManager.getAnonRole()==true){
+	 * Does so in a single query (unless anonymous mappings are missing, in which case they are created)
+	 * @param messages list of Messages
+	 */
+	private Map<String, String> getUserIdAnonIdMapForMessages(List messages)
```
add generics on "List messages"

bbailla2 replied:
Will do. Didn't know you can pass regular Lists into methods that take Lists with generics!

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -3381,6 +3405,73 @@ else if(getUserId()==null&&this.forumManager.getAnonRole()==true){
+		List<String> userIds = new ArrayList<String>();
+		for (Object objMessage : messages)
+		{
+			Message message = (Message) objMessage;
```
see above, casting should be unnecessary

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -4904,9 +4994,18 @@ public String processDfMsgSaveRevisedDraft()
-	          revisedInfo += getUserNameOrEid();
-	          revisedInfo += " " + getResourceBundleString(LAST_REVISE_ON);
+		      String revisedInfo;
+		      if (selectedTopic.getTopic().getPostAnonymous())
```
this chunk looks very similar to above logic. extract to private method.

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -8726,15 +8892,14 @@ private DiscussionForumBean duplicateForum(Long originalForumId) {
+		{
+			// Have to use getMessageByIdWithAttachments, or we'll get LazyInitializationExceptions
+			msg = messageManager.getMessageByIdWithAttachments(inReplyTo.getId());
+			inReplyTo = msg.getInReplyTo();
```
need null check here on msg?

bbailla2 replied:
If msg was null it would suggest that the previous line couldn't find a message with inReplyTo.getId(), which suggests that the reply hierarchy is broken. Messages are soft deleted though, so somebody would have had to tamper with the DB. I think it's okay to break hard in this situation, though we could stop what the user is doing and say "This message has been deleted" (but I'd argue that's out of scope)

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ForumsEmailService.java:
```diff
@@ -169,7 +172,7 @@ public void send() {
 			content.append(DiscussionForumTool
 					.getResourceBundleString("email.body.author")
-					+ " " + reply.getAuthor());
+					+ " " + anonAwareAuthor);
```
For consistency this string concat should be append() calls

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionMessageBean.java:
```diff
@@ -449,4 +457,128 @@ public String getAuthorEid(){
+
+	private boolean isAnonymousEnabled()
+	{
+		return ServerConfigurationService.getBoolean("msgcntr.forums.anonymous.enable", true);
```
As Brian said above, probably something like AnonymousManager.SAK_PROP_ANON_ENABLE is better here

bbailla2 replied:
Refactor will introduce AnonymousManager.isAnonymousEnabled()

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionMessageBean.java:
```diff
@@ -449,4 +457,128 @@ public String getAuthorEid(){
+			catch (RuntimeException e)
+			{
+				// Only act on this exception if its a hibernate LazyInitializationException (without depending on Hibernate as the ORM)
+				if ("org.hibernate.LazyInitializationException".equals(e.getClass().getName()))
```
Wondering if you should just attach the topic here, then check again, letting any runtime exception go unhandled for the second attempt. It would cover more cases and not rely on this magic string check.

bbailla2 replied:
The 3 lines below will always be executed within this catch (Ie. there will be no if / else)

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionMessageBean.java:
```diff
@@ -449,4 +457,128 @@ public String getAuthorEid(){
+				// Are we supposed to reveal authors' identities to certain roles in this topic?
+				if (topic.getRevealIDsToRoles())
+				{
+					UIPermissionsManager uiPermissionsManager = (UIPermissionsManager) ComponentManager.get("org.sakaiproject.api.app.messageforums.ui.UIPermissionsManager");
```
Maybe centralize all these ComponentManager dependency calls in AnonymousManager with methods like getUIPermissionsManager()? Easier to read and maintain instead of having them scattered around the code.

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionTopicBean.java:
```diff
@@ -556,6 +558,62 @@ public void setPostFirst(String postFirst)
+    LOG.debug("getRevealIDsToRoles()");
+    if ("".equals(revealIDsToRoles))
+    {
+      boolean isRevealIDsToRoles = topic != null && topic.getRevealIDsToRoles() != null && topic.getRevealIDsToRoles().booleanValue();
```
I think Booleans are just unboxed automatically now so the calls to booleanValue() aren't needed.

plukasew commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionTopicBean.java:
```diff
@@ -84,6 +84,8 @@
   private String locked = "";
   private String moderated = "";
   private String postFirst = "";
+  private String postAnonymous = "";
```
Why are these strings? They appear to be used as booleans later on.

bbailla2 replied:
I was just being consistent. It's possible something we're doing in the JSP files requires them to be Strings, although I think it's unlikely

plukasew commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/AnonymousManagerImpl.java:
```diff
@@ -0,0 +1,198 @@
+					if (e.getCause() instanceof ConstraintViolationException)
+					{
+						// Race condition!
+						LOG.warn("getOrCreateUserIdAnonIdMap: constraint violation while creating anonId for the following siteId, userId pair: (" + siteId + ", "  + userId + ")");
```
As you say, constraint violations are unavoidable here, I would drop this to info because there really isn't any action that can be taken here by anyone reading the logs.

plukasew commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java:
```diff
@@ -1547,6 +1562,57 @@ public Object doInHibernate(Session session) throws HibernateException, SQLExcep
+
+			List<Topic> topicList = new ArrayList<Topic>();
+			List resultSet = (List) getHibernateTemplate().execute(hcb);
+			for (Object objResultArray : resultSet)
```
Are we getting a mix of different objects here, or are they all going to be Topics?

bbailla2 replied:
IIRC I was getting a List of Object arrays, and each array contained [Topic, Forum, Area]. I think Hibernate sees the joins and includes it in the result set.

plukasew commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/UIPermissionsManagerImpl.java:
```diff
@@ -940,6 +941,34 @@ public boolean isModeratePostings(Long topicId, Boolean isForumLocked, Boolean i
+    }
+    catch (Exception e)
+    {
+      LOG.error(e.getMessage(), e);
```
What kind of exception are we going to get here? Should it be logged at error level?

bbailla2 replied:
I'm not seeing any obvious exceptions that could be thrown within this block. Just another instance copying the surrounding code.

plukasew commented on msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/UIPermissionsManagerImpl.java:
```diff
@@ -940,6 +941,34 @@ public boolean isModeratePostings(Long topicId, Boolean isForumLocked, Boolean i
+      while (iter.hasNext())
+      {
+        DBMembershipItem item = (DBMembershipItem) iter.next();
+        if (item.getPermissionLevel().getIdentifyAnonAuthors().booleanValue())
```
Automatic unboxing should take care of this so you don't need booleanValue()

plukasew commented on msgcntr/messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/PermissionLevelImpl.java:
```diff
@@ -195,6 +206,7 @@ public String toString() {
 		buffer.append("," + reviseAny);
 		buffer.append("," + reviseOwn);
 		buffer.append("," + moderatePostings);
+		buffer.append("," + identifyAnonAuthors);
```
string concat inside append() call, probably should just fix all of these

bbailla2 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ForumsEmailService.java:
```diff
@@ -345,4 +348,30 @@ private boolean isValidEmail(String email) {
+		if (isUseAnonymousId(message.getTopic()))
+		{
+			AnonymousManager anonMan = (AnonymousManager)ComponentManager.get("org.sakaiproject.api.app.messageforums.AnonymousManager");
+			String siteId = ToolManager.getCurrentPlacement().getContext();
```
ToolManager cover, should use API

bbailla2 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionMessageBean.java:
```diff
@@ -449,4 +457,128 @@ public String getAuthorEid(){
+	 */
+	public boolean isCurrentUserAndAnonymous()
+	{
+		return isUseAnonymousId() && message.getAuthorId().equals(SessionManager.getCurrentSessionUserId());
```
SessionManager is another cover

bbailla2 commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/MessageForumStatisticsBean.java:
```diff
@@ -2797,4 +3039,35 @@ public void setName(String name) {
+
+	private boolean isAnonymousEnabled()
+	{
+		return ServerConfigurationService.getBoolean("msgcntr.forums.anonymous.enable", true);
```
Centralize this

bbailla2 commented:
Committed refactor

buckett commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java:
```diff
@@ -3381,6 +3405,72 @@ else if(getUserId()==null&&this.forumManager.getAnonRole()==true){
+			return false;
+		}
+
+		// if topic.postAnonymous
```
Should this commented out code still be in here?

bbailla2 replied:
The return statement below was hard to wrap my head around, so I added this pseudocode for clarity.
I think it would be useful for anybody who needs to modify / understand this in the future.

buckett replied:
Ah, sorry didn't see the end of it. I'm all in favour of more readable code and agree it's hard to read. I'd write it out more fully in Java so that it's more self documenting.

    boolean revealAnonIds = topic.getRevealIDsToRoles() && uiPermissionsManager.isIdentifyAnonAuthors(topic);
    return topic.getPostAnonymous() && ! revealAnonIds;

That way it's easier if anyone ever has to debug it and doesn't have to trust that the comments and code are in sync. Although having written that I'm not sure my suggestion is much clearer.

buckett replied:
If you're ever looking at re-arranging logic Wolfram Alpha can be helpful, allowing you to enter the boolean expression and see simplified forms:

https://www.wolframalpha.com/input/?i=P+%26%26+%28%21Q+%7C%7C+%21R%29

bbailla2 replied:
^ Nice find!
I still have to say I prefer the current code, as it returns early when postAnonymous is false (the common case)

buckett commented:
Good cleanup of PermissionsMask and creation of  `setupMembershipItemPermission(DBMembershipItem, PermissionBean)`

buckett commented:
I've had a quick scan through and it's good, well done.

Nice to see the ToolPlacement stuff only in the tool and the SiteID exposed in the API. I think there are enitybroker endpoints for forums and this might allow someone to poke the API and find out who an anonymous user ID is (although I haven't double checked this). Also I think `org.sakaiproject.component.app.messageforums.entity.MessageForumsEntityContentProducer` will index the content and I don't think this is aware of anonymous topics.

From a design point of view it might have been cleaner to have all the anonymous IDs behind the API so that you can just retrieve the Topics and Messages and depending on the user and the topic the calls to `Message.getAuthor()` returns the anonymous ID when it's an anonymous topic and the normal username when it's not. This way the API acts as a choke on the data and it's much harder for a new feature to accidentally leak the read IDs in an anonymous topic. With all the work done this is coming a little late and also with the existing design/implementation of forums changes like these aren't always easy.

bbailla2 commented:
Thanks! 

I had to confirm, but you cannot entity link to messages; just forums and topics. 

I actually put 'getAnonAwareAuthor()' inside the Message API, but I didn't like it and I refactored it out. MessageImpl (a simple POJO with one line getters and setters) was getting bloated with ComponentManager calls and anonymous logic. The jsp files had references like #{selectedMessage.message.anonAwareAuthor} (as opposed to selectedMessage.anonAwareAuthor), and the Messages tool's implementation needed method stubs etc. So I refactored it all into DiscussionMessageBean, and it made everything a lot cleaner. But I see the concern that it would be easy to miss if messages were ever to be exposed outside of forums.

baholladay commented on msgcntr/messageforums-api/src/java/org/sakaiproject/api/app/messageforums/AnonymousMapping.java:
```diff
@@ -0,0 +1,22 @@
+public interface AnonymousMapping
+{
+	public Long getId();
+	public void setId(Long id);
```
This ID is probably not needed since UserId + SiteId will be unique

baholladay commented on msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/MessageForumStatisticsBean.java:
```diff
@@ -1971,22 +2167,24 @@ private String getExternalParameterByKey(String parameterId)
-		String lastName;
-		int firstIndex = 0;
-		int secondIndex = 0;
-		if (ServerConfigurationService.getBoolean("msg.displayEid", true)) {
```
Is this sakai.property feature removed and not added back?

bbailla2 replied:
Now that I'm looking at this again, both conditions are trying to parse the name into firstName + " " + lastName; the sakai.property was used to determine how to parse the 'selectedSiteUser' String to achieve this. The new code doesn't parse this property at all, and instead uses User.getDisplayName(). I'd argue that User.getDisplayName() is more appropriate than parsing the selectedSiteUser to generate firstName + " " + lastName (Ie. this considers contextualUserDisplayService, the DisplayAdvisoryUDP, etc.). So I think we're good here

baholladay commented on msgcntr/messageforums-hbm/src/java/org/sakaiproject/component/app/messageforums/dao/hibernate/AnonymousMapping.hbm.xml:
```diff
@@ -0,0 +1,25 @@
+<!DOCTYPE hibernate-mapping PUBLIC "-//Hibernate/Hibernate Mapping DTD 3.0//EN" "http://www.hibernate.org/dtd/hibernate-mapping-3.0.dtd">
+<hibernate-mapping>
+	<class name="org.sakaiproject.component.app.messageforums.dao.hibernate.AnonymousMappingImpl" table="MFR_ANONYMOUS_MAPPING_T">
+		<id column="ID" name="id" type="long">
```
Check whether ID is required, if not, I suggest removing it and relying on SITE_ID + USER_ID as the unique ID. Also, make sure that there is an index created for the SITE_ID + USER_ID combination

baholladay commented:
Completed a code review w/Brian

1) Verify/QA that emails sent to students who are "watching" the topic are anonymous 
2) Look into removing the auto ID in the DB and just use SITE_ID + USER_ID as the ID
3) Verify that the grade sort bug in Statistics and Grading by topic + user is not a regression
4) re-add the removed feature for "msg.displayEid"
5) Discuss with list about:
a) Should anonymous topic be able to be turned back to normal after any messages are posted?
b) Should the anonymous option be added to the Forum template?

My 2 cents:
5a) No, once a message has been posted, this feature should be locked.
5b) No, I agree w/Brian that the instructor could think that the topic is anonymous when it isn't.

bbailla2 commented:
1) They're in fact anonymous
2) Our DBA agreed it's unnecessary; I will remove the ID field
3) Confirmed that the issue is reproducible on nightly
4) Will do!
5) Only Adam Marshall weighed in. We're in agreement that we:
    a) should lock the option down
    b) should not include anonymity options in the forum templates

In terms of 5 a), I would opt to lock the option down immediately. I know some internet forums are locked for editing / deletion once a user posts to them, but we should prevent the following condition:
Instructor creates anonymous topic
Student begins composing, is informed their post will be anonymous
Instructor flips anonymity off
Student posts; all is revealed.

baholladay commented:
Thanks! 

5a) Yea, it makes sense to lock that option after its set and there is 1 post

bbailla2 commented:
Amended as per Bryan Holladay's suggestions
